samples/drivers/espi/src/espi_oob_handler.c

```diff
@@ -0,0 +1,214 @@
+
+	oob_hdr.dest_slave_addr = PCH_DEST_SLV_ADDR;
+	oob_hdr.oob_cmd_code = OOB_CMDCODE;
+	oob_hdr.byte_cnt = 1;
+	oob_hdr.src_slave_addr = SRC_SLV_ADDR;
+
+	/* Packetize OOB request */
+	req_pckt.buf = (uint8_t *)&oob_hdr;
+	req_pckt.len = sizeof(struct oob_header);
+
+	ret = espi_send_oob(dev, &req_pckt);
+	if (ret) {
+		LOG_ERR("OOB Tx failed %d", ret);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int retrieve_packet(const struct device *dev, uint8_t *sender)
+{
+	int ret;
+
+#ifdef CONFIG_ESPI_OOB_CHANNEL_RX_ASYNC
+	/* Note that no data is in the item */
+	uint8_t response_len;
+
+	if (k_msgq_num_used_get(&from_host) == 0U) {
+		return -EINVAL;
+	}
+
+	k_msgq_get(&from_host, &response_len, K_FOREVER);
+#endif
+
+	resp_pckt.buf = (uint8_t *)&buf;
+	resp_pckt.len = MAX_ESPI_BUF_LEN;
+
+	ret = espi_receive_oob(dev, &resp_pckt);
+	if (ret) {
+		LOG_ERR("OOB Rx failed %d", ret);
+		return ret;
+	}
+
+	LOG_INF("OOB transaction completed rcvd: %d bytes", resp_pckt.len);
+	for (int i = 0; i < resp_pckt.len; i++) {
+		LOG_INF("%x ", buf[i]);
+	}
+
+	if (sender) {
+		*sender = buf[OOB_RESPONSE_SENDER_INDEX];
+	}
+
+	return 0;
+}
+
+int get_pch_temp_sync(const struct device *dev)
+{
+	int ret;
+
+	for (int i = 0; i < MIN_GET_TEMP_CYCLES; i++) {
+		ret = request_temp(dev);
+		if (ret) {
+			LOG_ERR("OOB req failed %d", ret);
+			return ret;
+		}
+
+		ret = retrieve_packet(dev, NULL);
+		if (ret) {
+			LOG_ERR("OOB retrieve failed %d", ret);
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+int get_pch_temp_async(const struct device *dev)
+{
+#if !defined(CONFIG_ESPI_OOB_CHANNEL) || \
+	!defined(CONFIG_ESPI_OOB_CHANNEL_RX_ASYNC)
+	return -ENOTSUP;
+#else
+	context.espi_dev = dev;
+	context.cycles = MIN_GET_TEMP_CYCLES;
+
+	k_thread_start(espihub_thrd_id);
+	k_thread_join(espihub_thrd_id, K_FOREVER);
+
+	return 0;
+#endif
+}
+
+#ifdef CONFIG_ESPI_OOB_CHANNEL_RX_ASYNC
+
+void oob_rx_handler(const struct device *dev, struct espi_callback *cb,
+		    struct espi_event event)
+{
+	uint8_t last_resp_len = event.evt_details;
+
+	LOG_WRN("%s", __func__);
+	/* Post for post-processing in a thread
+	 * Should not attempt to retrieve in callback context
+	 */
+	k_msgq_put(&from_host, &last_resp_len, K_NO_WAIT);
+}
+
+
+bool need_temp;
+
+void temperature_timer(struct k_timer *timer_id)
+{
+	LOG_WRN("%s", __func__);
+	need_temp = true;
+}
+
+void espihub_thread(void *p1, void *p2, void *p3)
+{
+	int ret;
+	uint8_t temp;
+	uint8_t sender;
+
+	LOG_DBG("%s", __func__);
+	k_timer_start(&temp_timer, K_MSEC(100), K_MSEC(100));
+	while (context.cycles > 0) {
+		k_msleep(50);
+
+		ret = retrieve_packet(context.espi_dev, &sender);
+		if (!ret) {
+			switch (sender) {
+			case PCH_DEST_SLV_ADDR:
+				LOG_INF("PCH response");
+				/* Any other checks */
+				if (resp_pckt.len == OOB_RESPONSE_LEN) {
+					temp = buf[OOB_RESPONSE_DATA_INDEX];
+					LOG_INF("Temp %d", temp);
+				} else {
+					LOG_ERR("Incorrect size response");
+				}
+
+				context.cycles--;
```

Comment:
cycles is decremented only if sender == PCH_DEST_SLV_ADDR.
what is breaking condition if PCH hung or down and never responded to OOB req?

Reply:
Interesting question. If my memory is accurate the original eSPI specification did not provide maximum response times for upstream requests to OOB and Flash channels. You are correct the code above will only exit the thread if it obtains MIN_GET_TEMP_CYCLES (currently 5) temperatures.  Maybe the patch author could implement if N consecutive attempts fail then it exists the thread?

Reply:
Agreed, this will definitively cause this sample code to appear to be hung in any of above failing conditions. 
Code was just for reference and never intended as test case and it's not part of CI, but it can be corrected as @scottwcpg suggested.

Since this is sample code and backporting only, if everyone agrees we can track the improvement suggested here. https://github.com/zephyrproject-rtos/zephyr/issues/32457

Reply:
Sounds good.